### PR TITLE
[ROCm] Lowering ReduceOp with inline asm. Logical warpSize 32 is assumed.

### PR DESCRIPTION
**Pull request by weihanmines**
storeShared and shflSync implemented.

### CODE REVIEW

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 1]
```diff
@@ -392,12 +392,17 @@ static Value storeShared(ConversionPatternRewriter &rewriter, Location loc,
   unsigned bits = val.getType().getIntOrFloatBitWidth();
   const char *c = bits == 64 ? "l" : (bits == 16 ? "h" : "r");
 
+#if USE_ROCM
```
Comment (whchung) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:395]: @weihanmines `ctx`, `bits`, `c` are not used in `USE_ROCM` path either. The macro can be moved around to depict that.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 2]
```diff
@@ -1468,6 +1473,15 @@ Value ReduceOpConversion::shflSync(ConversionPatternRewriter &rewriter,
     return bitcast(vec, val.getType());
   }
 
+#ifdef USE_ROCM
+  DenseMap<short, unsigned int> masks{{16, 0x401F}, {8, 0x201F}, {4, 0x101F}, {2, 0x081F}, {1, 0x041F}};
```
Comment (whchung) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1477]: Please add comments on the meaning of the mask values.

**Review by whchung — CHANGES_REQUESTED** [review]
The use of inline asm needs to be revised. We issue 2 instructions, so please use 2 `builder.create` invocations.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 3]
```diff
@@ -1468,6 +1473,15 @@ Value ReduceOpConversion::shflSync(ConversionPatternRewriter &rewriter,
...
+  GCNBuilder builder; 
+  auto shfl = builder.create("ds_swizzle_b32");
+  auto dOpr = builder.newOperand("=v");
+  auto aOpr = builder.newOperand(val, "v");
+  auto maskOpr = builder.newConstantOperand("offset:" + std::to_string(masks[i])+"\ns_waitcnt lgkmcnt(0)\n");
```
[the remark this answers is not in the document]
Comment (whchung) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1482]: Please refer to https://github.com/ROCmSoftwarePlatform/triton/blob/triton-mlir-phase-2/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp#L4913-L4922 for how to insert multi-line inline asm.

**Comment by whchung** [comment]
@weihanmines please help revise the PR description so we remember currently the wavefront size is considered as 32. Therefore only half of the wavefront is participating in the `ds_swizzle`. We may need to revisit this decision when integrating `tt.reduce` op with other ops.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 1 above)
Comment (weihanmines) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:391]: moved bits and char pointer c to the end of storeShared Op in the latest PR.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 2 above)
Comment (weihanmines) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1477]: An explanation is added.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (weihanmines) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp]: Two instructions are inserted.

**Review by whchung — DISMISSED** [review]

**Comment by whchung** [comment]
@weihanmines Please add passing unit tests inside "test_core.py" into "test_core_amd.py" so it can be tracked by CI. After that we can merge this PR.

**Comment by weihanmines** [comment]
> @weihanmines Please add passing unit tests inside "test_core.py" into "test_core_amd.py" so it can be tracked by CI. After that we can merge this PR.

Unit tests regarding to reduce have been added.

**Review by whchung — APPROVED** [review]

**Comment by weihanmines** [comment]
> @weihanmines Please add passing unit tests inside "test_core.py" into "test_core_amd.py" so it can be tracked by CI. After that we can merge this PR.

Unit tests regarding to reduce have been added.

**Pull request merged**